File: src/js/gmemujs.js
!function (name, definition) {
    if (typeof module != 'undefined') module.exports = definition()
    else if (typeof define == 'function' && define.amd) define(name, definition)
    else this[name] = definition()
}('gmemujs', function () {
  
      var _Module = function () {
        include "../../tmp/build/a.out.js"
        return Module;
      }();

      var Module = function(Module) {
        return {
          initialize: Module.cwrap('initialize', "number", ["number", "number"]),
          openData: Module.cwrap('open_data', "number", ["number", "array", "number"]),
          trackCount: Module.cwrap('track_count', "number", ["number"]),
          openTrack: Module.cwrap('open_track', "number", ["number", "number"]),
          trackInfo: Module.cwrap('track_info', "string", ["number"]),
          trackStart: Module.cwrap('track_start', "number", ["number"]),
          generateSoundData: Module.cwrap('generate_sound_data', "number", ["number"]),
          getValue: Module.getValue.bind(Module)
        };
      }(_Module);
      return function(_Module) {

      var INT16_MAX = Math.pow(2, 32) - 1;
      var AudioContext = window.AudioContext || window.webkitAudioContext || window.mozAudioContext;
      var audioContext = new AudioContext();
      var scriptProcessor = audioContext.createScriptProcessor(8192, 2, 2);
      var bufferNode = audioContext.createBufferSource();
      var buffer = audioContext.createBuffer(2, 8192, audioContext.sampleRate);
      var gain = audioContext.createGain();
      var paused = false;


      var _c_albumBuilder = Module.initialize(44100, 8192);


      var gme = function() {

      };

      gme.Module = Module;

      gme.prototype = {};

      gme.load = function(path, cb) {
        var xhr = new XMLHttpRequest();
        xhr.open("GET", path, true);
        xhr.responseType = "arraybuffer";
        xhr.onload = function(e) {
          cb(new File(_c_albumBuilder, this.response), e, xhr);
        };
        xhr.send();
        return xhr;
      };

      gme.play = function(track) {
        Module.trackStart(track._c_track);
        scriptProcessor.onaudioprocess = function(e) {
          var bufferSize = 8192;
          var soundBuffer, channels;
          if (!paused) {
            soundBuffer = Module.generateSoundData(bufferSize);
            channels = [e.outputBuffer.getChannelData(0), e.outputBuffer.getChannelData(1)];

          }
          for (var i = 0; i < bufferSize; i++) {
            for (var n = 0; n < e.outputBuffer.numberOfChannels; n++) {
              var value = 0;
              if (!paused) {
                value = Module.getValue(soundBuffer + i * e.outputBuffer.numberOfChannels * 2 + n * 4, "i32") / INT16_MAX;
              }
              channels[n][i] = value;
            }
          }
        };
        gain.gain.value = 1;
        var data = buffer.getChannelData(0);
        for (var i = 0; i < 8192; i++) {
          data[i] = (Math.random() * 2) - 1;
        }
        data = buffer.getChannelData(1);
        for (i = 0; i < 8192; i++) {
          data[i] = (Math.random() * 2) - 1;
        }
        bufferNode.buffer = buffer;
        bufferNode.loop = true;
        bufferNode.connect(scriptProcessor);
        scriptProcessor.connect(gain);
        gain.connect(audioContext.destination);
        bufferNode.noteOn(0);
      };

      var File = function(albumBuilder, response) {
        var data = new Uint8Array(response);
        this._c_file = Module.openData(_c_albumBuilder, data, data.length);
      };

      File.prototype = {
        _c_file: undefined,
        track: function(trackNumber) {
          if (trackNumber > this.trackCount() || trackNumber < 0) {
            // do something
            return undefined;
          }

          return new Track(this, trackNumber);
        },
        _trackCount: undefined,
        trackCount: function() {
          return this._trackCount || (this._trackCount = Module.trackCount(this._c_file));
        },
      };

      var Track = function(file, trackNumber) {
        this.file = file;
        this.trackNumber = trackNumber;
        this._c_track = Module.openTrack(file._c_file, trackNumber);
      };

      Track.prototype = {
        _c_track: undefined,
        _info: undefined,
        info: function(field) {
          this._info = this._info || JSON.parse(Module.trackInfo(this._c_track));
          return field ? this._info[field] : {
            "length": this._info.length,
            "play_length": this._info.play_length, 
            "intro_length": this._info.intro_length, 
            "loop_length": this._info.loop_length,
            "system": this._info.system,
            "game": this._info.game,
            "song": this._info.song,
            "author": this._info.author,
            "copyright": this._info.copyright,
            "comment": this._info.comment,
            "dumper": this._info.dumper
          };
        }
      };

      return gme;
    }(Module);
 

});

File: src/c/gmemujs.c
#include <stdlib.h>
#include <stdio.h>
#include "gme.h"

const char* info_fmt = "{\"length\": %d, \"play_length\": %d, \"intro_length\": %d, \"loop_length\": %d, \"system\": \"%s\", \"game\": \"%s\", \"song\": \"%s\", \"author\": \"%s\", \"copyright\": \"%s\", \"comment\": \"%s\", \"dumper\": \"%s\"}";

static char json_str[2048];
static short audio_buffer[8192 * 2];
static Music_Emu* emu;
typedef struct AlbumBuilder
{
    int sample_rate;
    int buffer_size;
} AlbumBuilder;

typedef struct Album {
  AlbumBuilder* album_builder;
    int track_count;
} Album;

typedef struct Track {
    Album* album;
    int number;
} Track;

typedef struct PlayInfo
{
    AlbumBuilder * album_builder;
    Track * track;
} PlayInfo;
char* gmemujs_test () {
  return "hello world!";
}

AlbumBuilder * initialize (int sample_rate, int buffer_size) {
  AlbumBuilder * builder;
  builder = malloc(sizeof(AlbumBuilder));
  builder->sample_rate = sample_rate;
  builder->buffer_size = buffer_size;
  return builder;
}

Album * open_data (AlbumBuilder * builder, void const * data, long size) {
  Album * album;
  gme_open_data(data, size, &emu, builder->sample_rate);
  album = malloc(sizeof(Album));
  album->album_builder = builder;
  album->track_count = gme_track_count(emu);
  return album;
}

Track * open_track (Album * album, int track_number) {
  Track * track;
  track = malloc(sizeof(Track));
  track->album = album;
  track->number = track_number;
  return track;
}

char* track_info (Track * track) {
  gme_info_t * track_info;
  gme_track_info(emu, &track_info, track->number);
  sprintf(json_str, info_fmt, track_info->length,
    track_info->play_length, track_info->intro_length, track_info->loop_length,
    track_info->system, track_info->game, track_info->song, track_info->author,
    track_info->copyright, track_info->comment, track_info->dumper,
    track_info->loop_length, track_info->intro_length);
  return json_str;
}

int track_count (Album * album) {
  return album->track_count;
}


PlayInfo * play_info (Track * track) {
  PlayInfo * playinfo = malloc(sizeof(PlayInfo));
  playinfo->album_builder = track->album->album_builder;
  playinfo->track = track;
  return playinfo;
}

PlayInfo * track_start (Track * track) {
  PlayInfo * playinfo = play_info(track);
  gme_start_track(emu, track->number);
  return playinfo;
}

short * generate_sound_data (int buffer_size) {
  gme_play(emu, 8192 * 2, audio_buffer);
  return audio_buffer;
}
